src/blocks/scratch3_event.js

```diff
@@ -12,23 +12,76 @@ function Scratch3EventBlocks(runtime) {
  */
 Scratch3EventBlocks.prototype.getPrimitives = function() {
     return {
-        'event_whenflagclicked': this.whenFlagClicked,
-        'event_whenbroadcastreceived': this.whenBroadcastReceived,
-        'event_broadcast': this.broadcast
+        'event_broadcast': this.broadcast,
+        'event_broadcastandwait': this.broadcastAndWait,
+        'event_whengreaterthan': this.hatGreaterThanPredicate
     };
 };
 
+Scratch3EventBlocks.prototype.getHats = function () {
+    return {
+        'event_whenflagclicked': {
+            restartExistingThreads: true
+        },
+        /*'event_whenkeypressed': {
+            restartExistingThreads: false
+        },
+        'event_whenthisspriteclicked': {
+            restartExistingThreads: true
+        },
+        'event_whenbackdropswitchesto': {
+            restartExistingThreads: true
+        },*/
+        'event_whengreaterthan': {
+            restartExistingThreads: false,
+            edgeTriggered: true
+        },
+        'event_whenbroadcastreceived': {
+            restartExistingThreads: true
+        }
+    };
+};
 
-Scratch3EventBlocks.prototype.whenFlagClicked = function() {
-    // No-op
+Scratch3EventBlocks.prototype.hatGreaterThanPredicate = function (args, util) {
+    // @todo: Other cases :)
+    if (args.WHENGREATERTHANMENU == 'TIMER') {
+        return util.ioQuery('clock', 'projectTimer') > args.VALUE;
+    }
+    return false;
 };
 
-Scratch3EventBlocks.prototype.whenBroadcastReceived = function() {
-    // No-op
+Scratch3EventBlocks.prototype.broadcast = function(args, util) {
+    util.triggerHats('event_whenbroadcastreceived', {
+        'BROADCAST_OPTION': args.BROADCAST_OPTION
+    });
 };
 
-Scratch3EventBlocks.prototype.broadcast = function() {
-    // @todo
+Scratch3EventBlocks.prototype.broadcastAndWait = function (args, util) {
+    // Have we run before, triggering threads?
+    if (!util.stackFrame.triggeredThreads) {
+        // No - trigger hats for this broadcast.
+        util.stackFrame.triggeredThreads = util.triggerHats(
+            'event_whenbroadcastreceived', {
+                'BROADCAST_OPTION': args.BROADCAST_OPTION
+            }
+        );
+        if (util.stackFrame.triggeredThreads.length == 0) {
+            // Nothing was started.
+            return;
+        }
+    }
+    // We've run before; check if the wait is still going on.
+    var waiting = false;
+    for (var i = 0; i < util.stackFrame.triggeredThreads.length; i++) {
+        var thread = util.stackFrame.triggeredThreads[i];
+        var activeThreads = this.runtime.threads;
+        if (activeThreads.indexOf(thread) > -1) { // @todo: A cleaner way?
+            waiting = true;
+        }
+    }
```

Comment:
A few disconnected thoughts:
- I'm OK with control structures acting as a special class of blocks which get a little extra access to internals. Here I'm counting hat blocks as "control structures" in addition to if, repeat, etc.
- If we do want extensions to be able to do this sort of thing, I'd feel better about exposing methods on `runtime` to extensions than the `threads` array itself. Maybe something like `runtime.isActiveThread(thread)` which returns a Boolean?
- I've seen some examples of `for (var i = 0; arrayOfObjects[i]; ++i) { doSomethingWith(arrayOfObjects[i]); }` in a few places where the array is known to contain non-false-like things at each index until the end. I think of this as "the Google way" but I can't remember why. Anyway, allegedly it can be faster than querying `length` each time, but it's also pinging my "premature optimization" detector a bit. What do you think?
- I think you can actually reduce this even further using something like the below. Actually, now that I've written it this way, I think this might be more readable too.
```js
waiting = util.stackFrame.triggeredThreads.some(function(thread) {
    return this.runtime.threads.indexOf(thread) >= 0;
};
// or
waiting = util.stackFrame.triggeredThreads.some(function(thread) {
    return this.runtime.isActiveThread(thread);
};
```
For reference: [`Array.prototype.some()`](http://kangax.github.io/compat-table/es5/#test-Array_methods_Array.prototype.some) (you may have to expand the "Array methods" section).